File: companies/random/centre_tree.java
import java.util.*;

class Main {
    public static void main(String[] args) {
        List<int[]> edges1 = new ArrayList<>();
        edges1.add(new int[]{1, 2});
        edges1.add(new int[]{2, 3});
        edges1.add(new int[]{3, 4});
        edges1.add(new int[]{3, 5});
        edges1.add(new int[]{1, 6});
        edges1.add(new int[]{1, 7});

        int[] result1 = solve(edges1);
        System.out.println(Arrays.toString(result1));

        List<int[]> edges2 = new ArrayList<>();
        edges2.add(new int[]{2, 1});

        int[] result2 = solve(edges2);
        System.out.println(Arrays.toString(result2));
    }

    static class Edge {
        int v;
        int w;

        Edge(int v, int w) {
            this.v = v;
            this.w = w;
        }
    }

    public static int[] solve(List<int[]> edges) {
        int n = edges.size() + 1;
        List<List<Edge>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            int u = edge[0] - 1;
            int v = edge[1] - 1;
            adj.get(u).add(new Edge(v, 1));
            adj.get(v).add(new Edge(u, 1));
        }

        int[] d1 = dijkstra(0, adj);
        int v1 = -1;
        int maxDist1 = -1;

        for (int i = 0; i < n; i++) {
            if (d1[i] > maxDist1) {
                maxDist1 = d1[i];
                v1 = i;
            }
        }

        int[] d2 = dijkstra(v1, adj);
        int v2 = -1;
        int maxDist2 = -1;

        for (int i = 0; i < n; i++) {
            if (d2[i] > maxDist2) {
                maxDist2 = d2[i];
                v2 = i;
            }
        }

        List<Integer> path = new ArrayList<>();
        while (v2 != v1) {
            path.add(v2);
            v2 = par[v2];
        }
        path.add(v1);

        int[] res = new int[n];
        int mid = path.size() / 2;

        if (path.size() % 2 == 0) {
            int c1 = path.get(mid - 1);
            int c2 = path.get(mid);
            for (int v : leaves(c1, c2, adj)) {
                res[v] = 1;
            }
            for (int v : leaves(c2, c1, adj)) {
                res[v] = 1;
            }
        } else {
            int c1 = path.get(mid);
            for (int v : leaves(c1, c1, adj)) {
                res[v] = 1;
            }
        }

        return res;
    }

    static int[] par;

    public static int[] dijkstra(int src, List<List<Edge>> adj) {
        int n = adj.size();
        int[] dist = new int[n];
        par = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        Arrays.fill(par, -1);

        PriorityQueue<Edge> pq = new PriorityQueue<>(Comparator.comparingInt(e -> e.w));
        pq.add(new Edge(src, 0));
        dist[src] = 0;
        par[src] = src;

        while (!pq.isEmpty()) {
            Edge edge = pq.poll();
            int u = edge.v;
            int c = edge.w;

            if (c > dist[u]) {
                continue;
            }

            for (Edge neighbor : adj.get(u)) {
                int v = neighbor.v;
                int w = neighbor.w;
                if (dist[v] > dist[u] + w) {
                    dist[v] = dist[u] + w;
                    par[v] = u;
                    pq.add(new Edge(v, dist[v]));
                }
            }
        }

        return dist;
    }

    public static List<Integer> leaves(int u, int p, List<List<Edge>> adj) {
        Queue<Integer> q = new LinkedList<>();
        q.offer(u);
        List<Integer> ret = new ArrayList<>();
        Set<Integer> vis = new HashSet<>();
        vis.add(u);
        vis.add(p);

        while (!q.isEmpty()) {
            int sz = q.size();
            List<Integer> lvl = new ArrayList<>();

            for (int i = 0; i < sz; i++) {
                int v = q.poll();
                lvl.add(v);

                for (Edge edge : adj.get(v)) {
                    int neighbor = edge.v;
                    if (!vis.contains(neighbor)) {
                        q.offer(neighbor);
                        vis.add(neighbor);
                    }
                }
            }

            ret = lvl;
        }

        return ret;
    }
}

File: companies/random/centre_tree.py
from collections import deque
import heapq
def dijsktra(src, adj):
    n = len(adj)
    dist = [float('inf')]*n
    par = [i for i in range(n)]
    q = [(0, src)]
    dist[src] = 0
    par[src] = src
    while (q):
        c, u = heapq.heappop(q)
        if c > dist[u]:
            continue
        for v, w in adj[u]:
            if dist[v] > dist[u]+w:
                dist[v] = dist[u] + w
                par[v] = u;
                heapq.heappush(q, (dist[v], v))
    return dist,par

def leaves(u,p,adj):
    q= deque()
    q.append(u)
    ret = []
    vis = set([u,p])
    while q:
        lvl = []
        sz = len(q)
        for  i in range(sz):
            v = q.popleft()
            lvl.append(v)
            for u,w in adj[v]:
                if u not in vis: 
                    q.append(u)
                    vis.add(u)
        ret = lvl
    return ret
def solve(edges):
    n = len(edges) +1
    adj = [[] for i in range(n)]
    for u,v in edges:
        u -=1
        v -=1
        adj[u].append((v,1))
        adj[v].append((u,1))
    d,_ = dijsktra(0,adj)
    v1 = d.index(max(d))
    d2,par = dijsktra(v1,adj)
    v2 = d2.index(max(d2))
    path = [v2]
    while par[path[-1]]!=path[-1]:
        path.append(par[path[-1]])
    res = [0]*n
    mid = len(path)//2
    print(v1,v2,path)
    if len(path)%2==0:
        c1,c2 = path[mid-1],path[mid]
        for v in leaves(c1,c2,adj) + leaves(c2,c1,adj):
            res[v] = 1
    else:
        c1 = path[mid]
        for v in leaves(c1,c1,adj):
            res[v] = 1
        
    return res

print(solve([[1,2],[2,3],[3,4],[3,5],[1,6],[1,7]]))
print(solve([[2,1]] ))